elastic/beats
# Add ingress nginx controller fileset

Resolves: https://github.com/elastic/beats/issues/8451

This PR adds fileset for [`ingress-nginx`](https://kubernetes.github.io/ingress-nginx/) controller logs. Log patterns could be found on the controllers' [docs](https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/log-format.md).

~~In order to add the new pipeline a new setting called "version" is being introduced also in this PR. See the sum-up of the discussion on this [comment](https://github.com/elastic/beats/issues/8451#issuecomment-583387725).~~

![Screenshot 2020-02-11 at 16 44 03](https://user-images.githubusercontent.com/11754898/74246930-f2a3c280-4ced-11ea-8b98-0518cf92431f.png)


### How to test this

#### Filebeat
0. in `/tmp/testnginx` one can use the sample logs(see `test.log` file) used for automated testing. 
1. Enable nginx Filebeat module
2. Configure the fileset.
```
- module: nginx
  # Access logs
  ingress_controller:
    enabled: true
    var.paths: ["/tmp/testnginx"]
```
3. Setup the pipelines in ES: `./filebeat setup`
4. Start Filebeat and verify that events are properly shown in Kibana. NOTE: set the time window properly in Kibana so as to display the events in the timeline since the events have timestamps which will be in the past. 

#### k8s environment
1. Deploy an ingress managed sample app on Minikube following the official [k8s docs](https://kubernetes.io/docs/tasks/access-application-cluster/ingress-minikube/)
2. check that ingress controller logs are being populated when doing curls (like described in k8s docs). Ingress controller pod lies under `kube-system` namespace and you will be able to check its logs with sth like `nginx-ingress-controller-6fc5bcc8c9-zm8zv`.
3. Redirect pods' logs to a temporary file: `kubectl -n kube-system logs -f nginx-ingress-controller-6fc5bcc8c9-zm8zv >> /tmp/ingresspod `
4. Inside minikube configure nginx module like:
```
- module: nginx
  # Ingress-nginx controller logs. This is disabled by default. It could be used in Kubernetes environments to parse ingress-nginx logs
  ingress_controller:
    enabled: true

    # Set custom paths for the log files. If left empty,
    # Filebeat will choose the paths depending on your OS.
    var.paths: ["/tmp/ingresspod"]
``` 
5. Run filebeat and check that logs are parsed correctly and shown in Kibana properly.
6. Visit `http://hello-world.info/v2` so as to create new log entries.
